json_rpc/private/jrpc_sys.nim

```diff
@@ -0,0 +1,350 @@
+    r.parseArray:
+      val.positional.add ParamDescRx(
+        kind: r.tokKind(),
+        param: r.parseAsString(),
+      )
+  of JsonValueKind.Object:
+    val = RequestParamsRx(kind: rpNamed)
+    val.named = r.parseAsString()
+  else:
+    r.raiseUnexpectedValue("RequestParam must be either array or object, got=" & $tok)
+
+proc writeValue*[F](w: var JsonWriter[F], val: RequestParamsRx) =
+  {.error: "RequestParamsRx cannot be encoded, please use RequestParamsTx".}
+
+proc readValue*[F](r: var JsonReader[F], val: var RequestParamsTx) =
+  {.error: "RequestParamsTx cannot be decoded, please use RequestParamsRx".}
+
+proc writeValue*(w: var JsonWriter[JrpcSys], val: RequestParamsTx)
+      {.gcsafe, raises: [IOError].} =
+  case val.kind
+  of rpPositional:
+    w.writeArray val.positional
+  of rpNamed:
+    w.beginRecord RequestParamsTx
+    for x in val.named:
+      w.writeField(x.name, x.value)
+    w.endRecord()
+
+proc readValue*[F](r: var JsonReader[F], val: var RequestTx) =
+  {.error: "RequestTx cannot be decoded, please use RequestRx".}
+
+# Generated by json-serialization
+# proc writeValue*(w: var JsonWriter[JrpcSys], val: RequestTx)
+
+# Generated by json-serialization
+# proc readValue*(r: var JsonReader[JrpcSys], val: var RequestRx)
+
+proc writeValue*[F](w: var JsonWriter[F], val: RequestRx) =
+  {.error: "RequestRx cannot be encoded, please use RequestTx".}
+
+proc readValue*[F](r: var JsonReader[F], val: var ResponseTx) =
+  {.error: "ResponseTx cannot be decoded, please use ResponseRx".}
+
+proc writeValue*(w: var JsonWriter[JrpcSys], val: ResponseTx)
+       {.gcsafe, raises: [IOError].} =
+  w.beginRecord ResponseTx
+  w.writeField("jsonrpc", val.jsonrpc)
+  w.writeField("id", val.id)
+  if val.kind == rkResult:
+    w.writeField("result", val.result)
+  else:
+    w.writeField("error", val.error)
+  w.endRecord()
+
+# Generated by json-serialization
+# proc readValue*(r: var JsonReader[JrpcSys], val: var ResponseRx)
+
+proc writeValue*[F](w: var JsonWriter[F], val: ResponseRx) =
+  {.error: "ResponseRx cannot be encoded, please use ResponseTx".}
+
+proc readValue*[F](r: var JsonReader[F], val: var RequestBatchTx) =
+  {.error: "RequestBatchTx cannot be decoded, please use RequestBathRx".}
+
+proc writeValue*(w: var JsonWriter[JrpcSys], val: RequestBatchTx)
+       {.gcsafe, raises: [IOError].} =
+  if val.kind == rbkMany:
+    w.writeArray(val.many)
+  else:
+    w.writeValue(val.single)
+
+proc readValue*(r: var JsonReader[JrpcSys], val: var RequestBatchRx)
+       {.gcsafe, raises: [IOError, SerializationError].} =
+  let tok = r.tokKind
+  case tok
+  of JsonValueKind.Array:
+    val = RequestBatchRx(kind: rbkMany)
+    r.readValue(val.many)
+  of JsonValueKind.Object:
+    val = RequestBatchRx(kind: rbkSingle)
+    r.readValue(val.single)
+  else:
+    r.raiseUnexpectedValue("RequestBatch must be either array or object, got=" & $tok)
+
+proc writeValue*[F](w: var JsonWriter[F], val: RequestBatchRx) =
```

Comment:
ditto on object safety - it seems better to use an opt-in serializer here, also for future refactoring when other types that shouldn't be serialized accidentally might end up getting matched.

Reply:
fixed, they are now using custom serializer.